File: agents/examples/web-sdk-server/src/main/java/com/hmdev/messaging/common/util/EnvLoader.java
package com.hmdev.messaging.common.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

/**
 * Utility class to load environment variables from .env file.
 *
 * This loader supports the open-source SDK repo while allowing
 * secure configuration from the private services repo.
 *
 * Search Order:
 * 1. Current project root (.env)
 * 2. Services repo (../../messaging-platform-services/.env)
 * 3. Falls back to system properties/environment variables
 *
 * Property Names (consistent across all repos):
 * - MESSAGING_API_KEY: Developer API key for authentication
 * - DEFAULT_API_KEY: Default API key (legacy support)
 * - MESSAGING_API_URL: URL of messaging service API
 * - FILE_API_URL: URL of file service API
 * - ADMIN_EMAIL: Admin account email
 * - ADMIN_PASSWORD: Admin account password
 *
 * Note: Empty string values are treated as non-existent (skipped in favor of next source).
 */
public class EnvLoader {

    private static final Logger logger = LoggerFactory.getLogger(EnvLoader.class);

    private static final String ENV_FILE_NAME = ".env";
    private static final Map<String, String> envVariables = new HashMap<>();
    private static boolean loaded = false;

    // Built-in defaults (used only if no other configuration is found)
    private static final Map<String, String> DEFAULT_VALUES = new HashMap<>() {{
        put("MESSAGING_API_URL", "https://hmdevonline.com/messaging-platform/api/v1/messaging-service");
        // Note: API keys should be configured via .env or environment variables
    }};

    /**
     * Load .env file and update specified system properties
     * @param systemPropertyKeysToUpdate System property keys to update from loaded configuration
     */
    public static synchronized void load(String ...systemPropertyKeysToUpdate) {
        if (loaded) {
            return;
        }

        File envFile = findEnvFile();

        if (envFile != null && envFile.exists()) {
            logger.info("Loading environment from: {}", envFile.getAbsolutePath());
            loadFromFile(envFile);
        } else {
            logger.debug("No .env file found. Using system properties/environment variables.");
        }

        loaded = true;

        for (String key : systemPropertyKeysToUpdate) {
            // Set system properties from loaded configuration
            updateSystemProperty(key);
        }
    }

    /**
     * Find .env file in search order
     */
    private static File findEnvFile() {
        // 1. Try current project root
        File currentProjectEnv = findInCurrentProject();
        if (currentProjectEnv != null && currentProjectEnv.exists()) {
            return currentProjectEnv;
        }

        // 2. Try services repo (sibling to SDK repo)
        File servicesRepoEnv = findInServicesRepo();
        if (servicesRepoEnv != null && servicesRepoEnv.exists()) {
            return servicesRepoEnv;
        }

        return null;
    }

    /**
     * Find .env in current project root
     */
    private static File findInCurrentProject() {
        // Get current working directory
        String userDir = System.getProperty("user.dir");

        // Navigate up to find project root (look for settings.gradle or build.gradle)
        Path currentPath = Paths.get(userDir);

        while (currentPath != null) {
            // Check for .env in current directory
            File envFile = new File(currentPath.toFile(), ENV_FILE_NAME);
            if (envFile.exists()) {
                return envFile;
            }

            // Also check for .env in messaging-platform-services subdirectory
            File servicesSubdirEnv = new File(new File(currentPath.toFile(), "messaging-platform-services"), ENV_FILE_NAME);
            if (servicesSubdirEnv.exists()) {
                return servicesSubdirEnv;
            }

            // Check if this is project root (has settings.gradle or build.gradle)
            File settingsGradle = new File(currentPath.toFile(), "settings.gradle");
            File buildGradle = new File(currentPath.toFile(), "build.gradle");

            if (settingsGradle.exists() || buildGradle.exists()) {
                // This is project root, stop searching
                break;
            }

            currentPath = currentPath.getParent();
        }

        return null;
    }

    /**
     * Find .env in services repo (assuming SDK and services are siblings)
     */
    private static File findInServicesRepo() {
        try {
            // Get current working directory
            String userDir = System.getProperty("user.dir");
            Path currentPath = Paths.get(userDir);

            // Navigate up to find SDK root (contains "messaging-platform-sdk" in path)
            while (currentPath != null) {
                String dirName = currentPath.getFileName().toString();

                if (dirName.equals("messaging-platform-sdk")) {
                    // Found SDK root, now look for services repo as sibling
                    Path parentPath = currentPath.getParent();

                    if (parentPath != null) {
                        Path servicesPath = parentPath.resolve("messaging-platform-services");
                        File servicesEnvFile = servicesPath.resolve(ENV_FILE_NAME).toFile();

                        if (servicesEnvFile.exists()) {
                            return servicesEnvFile;
                        }
                    }

                    break;
                }

                // Also check for "messaging-platform-services" in path
                if (dirName.equals("messaging-platform-services")) {
                    // Found services root, look for .env in root
                    File servicesEnvFile = new File(currentPath.toFile(), ENV_FILE_NAME);
                    if (servicesEnvFile.exists()) {
                        return servicesEnvFile;
                    }
                    break;
                }

                currentPath = currentPath.getParent();
            }
        } catch (Exception e) {
            logger.warn("Error searching for services repo .env: {}", e.getMessage());
        }

        return null;
    }

    /**
     * Load variables from .env file
     */
    private static void loadFromFile(File envFile) {
        try (BufferedReader reader = new BufferedReader(new FileReader(envFile))) {
            String line;
            int lineNumber = 0;

            while ((line = reader.readLine()) != null) {
                lineNumber++;
                line = line.trim();

                // Skip empty lines and comments
                if (line.isEmpty() || line.startsWith("#")) {
                    continue;
                }

                // Parse KEY=VALUE
                int equalsIndex = line.indexOf('=');
                if (equalsIndex > 0) {
                    String key = line.substring(0, equalsIndex).trim();
                    String value = line.substring(equalsIndex + 1).trim();

                    // Remove quotes if present
                    if ((value.startsWith("\"") && value.endsWith("\"")) ||
                        (value.startsWith("'") && value.endsWith("'"))) {
                        value = value.substring(1, value.length() - 1);
                    }

                    envVariables.put(key, value);
                } else {
                    logger.warn("Invalid line {} in {}: {}", lineNumber, envFile.getName(), line);
                }
            }

            logger.debug("Loaded {} variables from .env file", envVariables.size());

        } catch (IOException e) {
            logger.error("Error reading .env file: {}", e.getMessage());
        }
    }

    /**
     * Check if a value is valid (non-null and non-empty)
     */
    private static boolean isValidValue(String value) {
        return value != null && !value.isEmpty();
    }

    /**
     * Get environment variable (from .env, system properties, or environment)
     */
    public static String get(String key) {
        return get(key, null);
    }

    /**
     * Get environment variable with default value
     */
    public static String get(String key, String defaultValue) {
        // Ensure .env is loaded
        load();

        // 1. Check loaded .env variables (direct key)
        String value = envVariables.get(key);
        if (isValidValue(value)) {
            return value;
        }

        // 2. Check system properties
        value = System.getProperty(key);
        if (isValidValue(value)) {
            return value;
        }

        // 3. Check environment variables
        value = System.getenv(key);
        if (isValidValue(value)) {
            return value;
        }

        // 4. Check built-in defaults
        value = DEFAULT_VALUES.get(key);
        if (isValidValue(value)) {
            return value;
        }

        // 5. Return provided default
        return defaultValue;
    }

    /**
     * Set system property from .env variable or built-in defaults
     * @param key Property key to set
     */
    public static void updateSystemProperty(String key) {
        String value = get(key);
        if (isValidValue(value)) {
            System.setProperty(key, value);
        }
    }

    /**
     * Set system property from .env variable with custom default
     * @param key Property key to set
     * @param defaultValue Default value if not found in any configuration source
     */
    public static void updateSystemProperty(String key, String defaultValue) {
        String value = get(key, defaultValue);
        if (isValidValue(value)) {
            System.setProperty(key, value);
        }
    }
}

File: agents/python-agent/hmdev/messaging/agent/util/env_loader.py
"""
Utility module to load environment variables from .env file.

This loader supports the open-source SDK repo while allowing
secure configuration from the private services repo.

Search Order:
1. Current project root (.env)
2. Services repo (../../messaging-platform-services/.env)
3. Falls back to os.environ environment variables

Property Names (consistent across all repos):
- MESSAGING_API_KEY: Developer API key for authentication
- DEFAULT_API_KEY: Default API key (legacy support)
- MESSAGING_API_URL: URL of messaging service API

Note: Empty string values are treated as non-existent (skipped in favor of next source).
"""

import os
import logging
from pathlib import Path
from typing import Optional, Dict

logger = logging.getLogger(__name__)

# Module-level state
_env_variables: Dict[str, str] = {}
_loaded: bool = False

# Built-in defaults (used only if no other configuration is found)
_DEFAULT_VALUES: Dict[str, str] = {
    "MESSAGING_API_URL": "https://hmdevonline.com/messaging-platform/api/v1/messaging-service",
    # Note: API keys should be configured via .env or environment variables
}

ENV_FILE_NAME = ".env"


def load() -> None:
    """
    Load .env file if not already loaded.
    Searches in order:
    1. Current project root
    2. Services repo (sibling to SDK repo)
    """
    global _loaded, _env_variables

    if _loaded:
        return

    env_file = _find_env_file()

    if env_file and env_file.exists():
        logger.info(f"Loading environment from: {env_file.absolute()}")
        _load_from_file(env_file)
    else:
        logger.debug("No .env file found. Using environment variables.")

    _loaded = True


def _find_env_file() -> Optional[Path]:
    """Find .env file in search order."""
    # 1. Try current project root
    current_project_env = _find_in_current_project()
    if current_project_env and current_project_env.exists():
        return current_project_env

    # 2. Try services repo (sibling to SDK repo)
    services_repo_env = _find_in_services_repo()
    if services_repo_env and services_repo_env.exists():
        return services_repo_env

    return None


def _find_in_current_project() -> Optional[Path]:
    """Find .env in current project root."""
    # Get current working directory
    current_path = Path.cwd()

    # Navigate up to find project root
    while current_path.parent != current_path:  # Stop at filesystem root
        # Check for .env in current directory
        env_file = current_path / ENV_FILE_NAME
        if env_file.exists():
            return env_file

        # Also check for .env in messaging-platform-services subdirectory
        services_subdir_env = current_path / "messaging-platform-services" / ENV_FILE_NAME
        if services_subdir_env.exists():
            return services_subdir_env

        # Check if this is project root (has common project indicators)
        if (current_path / "setup.py").exists() or \
           (current_path / "pyproject.toml").exists() or \
           (current_path / "build.gradle").exists() or \
           (current_path / "settings.gradle").exists():
            # This is project root, stop searching
            break

        current_path = current_path.parent

    return None


def _find_in_services_repo() -> Optional[Path]:
    """Find .env in services repo (assuming SDK and services are siblings)."""
    try:
        # Get current working directory
        current_path = Path.cwd()

        # Navigate up to find SDK root (contains "messaging-platform-sdk" in path)
        while current_path.parent != current_path:  # Stop at filesystem root
            if current_path.name == "messaging-platform-sdk":
                # Found SDK root, now look for services repo as sibling
                parent_path = current_path.parent
                services_path = parent_path / "messaging-platform-services"
                services_env_file = services_path / ENV_FILE_NAME

                if services_env_file.exists():
                    return services_env_file

                break

            current_path = current_path.parent
    except Exception as e:
        logger.warning(f"Error searching for services repo .env: {e}")

    return None


def _load_from_file(env_file: Path) -> None:
    """Load variables from .env file."""
    try:
        with open(env_file, 'r', encoding='utf-8') as f:
            line_number = 0
            for line in f:
                line_number += 1
                line = line.strip()

                # Skip empty lines and comments
                if not line or line.startswith('#'):
                    continue

                # Parse KEY=VALUE
                if '=' in line:
                    key, _, value = line.partition('=')
                    key = key.strip()
                    value = value.strip()

                    # Remove quotes if present
                    if (value.startswith('"') and value.endswith('"')) or \
                       (value.startswith("'") and value.endswith("'")):
                        value = value[1:-1]

                    _env_variables[key] = value
                else:
                    logger.warning(f"Invalid line {line_number} in {env_file.name}: {line}")

        logger.debug(f"Loaded {len(_env_variables)} variables from .env file")

    except IOError as e:
        logger.error(f"Error reading .env file: {e}")


def _is_valid_value(value: Optional[str]) -> bool:
    """Check if a value is valid (not None and not empty)."""
    return value is not None and value != ""


def get(key: str, default_value: Optional[str] = None) -> Optional[str]:
    """
    Get environment variable from multiple sources.

    Search order:
    1. Loaded .env variables
    2. os.environ environment variables
    3. Built-in defaults
    4. Provided default value

    Args:
        key: Environment variable key
        default_value: Default value if not found

    Returns:
        The environment variable value or default
    """
    # Ensure .env is loaded
    load()

    # 1. Check loaded .env variables
    value = _env_variables.get(key)
    if _is_valid_value(value):
        return value

    # 2. Check environment variables
    value = os.environ.get(key)
    if _is_valid_value(value):
        return value

    # 3. Check built-in defaults
    value = _DEFAULT_VALUES.get(key)
    if _is_valid_value(value):
        return value

    # 4. Return provided default
    return default_value
